File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ExecutorCache.scala
/*
 * Copyright (c) 2025, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids

import java.lang.management.ManagementFactory;

import ai.rapids.cudf.{Cuda, CudaComputeMode}

/**
 * A singleton object to cache executor related information.
 * Uses lazy mode to ensure the values are only computed once per executor.
 */
object ExecutorCache {

  /**
   * Cache the current device compute mode for current executor.
   * It's based on the assumption that executor has been assigned to a single device,
   * and will not change during the lifetime of the executor.
   * Should be called on executor side.
   */
  lazy val getCurrentDeviceComputeMode: CudaComputeMode = Cuda.getComputeMode()

  /**
   * Cache the current device UUID for current executor.
   * It's based on the assumption that executor has been assigned to a single device,
   * and will not change during the lifetime of the executor.
   * Should be called on executor side.
   */
  lazy val getCurrentDeviceUuid: Array[Byte] = Cuda.getGpuUuid()

  /**
   * Cache the current process name for current executor.
   * Should be called on executor side.
   */
  lazy val getProcessName: String = ManagementFactory.getRuntimeMXBean.getName
}

File: sql-plugin-columnar/src/main/java/com/nvidia/spark/rapids/ExecutorCache.java
/*
 * Copyright (c) 2025-2026, NVIDIA CORPORATION.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nvidia.spark.rapids;

import java.lang.management.ManagementFactory;

import ai.rapids.cudf.Cuda;
import ai.rapids.cudf.CudaComputeMode;

/**
 * Caches executor-related information. Values are initialized lazily to match the previous Scala
 * object semantics.
 */
final class ExecutorCache {
  private ExecutorCache() {
  }

  static CudaComputeMode getCurrentDeviceComputeMode() {
    return CurrentDeviceComputeModeHolder.VALUE;
  }

  static byte[] getCurrentDeviceUuid() {
    return CurrentDeviceUuidHolder.VALUE;
  }

  static String getProcessName() {
    return ProcessNameHolder.VALUE;
  }

  private static final class CurrentDeviceComputeModeHolder {
    private static final CudaComputeMode VALUE = Cuda.getComputeMode();
  }

  private static final class CurrentDeviceUuidHolder {
    private static final byte[] VALUE = Cuda.getGpuUuid();
  }

  private static final class ProcessNameHolder {
    private static final String VALUE = ManagementFactory.getRuntimeMXBean().getName();
  }
}
